programs/server/Server.cpp

```diff
@@ -483,16 +484,11 @@ int Server::main(const std::vector<std::string> & /*args*/)
         }
     }
 
-    if (config().has("interserver_http_credentials"))
-    {
-        String user = config().getString("interserver_http_credentials.user", "");
-        String password = config().getString("interserver_http_credentials.password", "");
-
-        if (user.empty())
-            throw Exception("Configuration parameter interserver_http_credentials user can't be empty", ErrorCodes::NO_ELEMENTS_IN_CONFIG);
-
-        global_context->setInterserverCredentials(user, password);
-    }
+    LOG_DEBUG(log, "Initiaializing InterserverCredentials.");
+    global_context->updateInterserverCredentials(config());
+    if (global_context->getInterserverCredential() == nullptr)
+        throw Exception("Error initializing interserver credentials",
+            ErrorCodes::NO_ELEMENTS_IN_CONFIG);
```

Comment:
How is this possible?

Reply:
Correct, the code is unreachable. It's there in case it changes in the future

Reply:
Same here, I'm suggesting removing it

Reply:
We normally use a C++ reference, not pointer, for things that can't be nullptr. Didn't read the code, but maybe it's suitable here as well?

If it *can* be nullptr in some cases, but shouldn't be nullptr here -- a suitable exception code would be `LOGICAL_ERROR`. It is meant for conditions that should never occur and signify a program error.

Reply:
Ok, so I should update the callee to throw (LOGICAL_ERROR)  in case of an error and catch the reception in Server.cpp?

Reply:
>We normally use a C++ reference, not pointer, for things that can't be nullptr. Didn't read the code, but maybe it's suitable here as well?

There is shared_ptr, indeed strange to see check against nullptr

But I actually suggesting remove this piece of code anyway.

Reply:
Server creates credentials when sever starts -- or throws and the server tattoos to start
When config is reloaded the server should never reset credentials with a nullptr. Existing code will never trigger the branch but future code might. Just trying to be safe here but since it does nothing it makes sense to remove

Reply:
> Ok, so I should update the callee to throw (LOGICAL_ERROR) in case of an error and catch the reception in Server.cpp?

First, determine when it is correct for `getInterserverCredential()` to return nullptr. If it is never correct, it should return a reference, not a pointer.

If it can return nullptr, the next question is whether it shares ownership of interserver credentials with the caller. If not, returning a raw pointer is more suitable.

The next question is whether you want to meaningfully handle the nullptr here, or is it always a program error. If it is a program error, you don't really have to add a special check for nullptr, the server is going to fail with a segfault. We don't consistently check for null pointers everywhere, so this strategy is almost acceptable. If you prefer an exception, it should have the code LOGICAL_ERROR. You don't have to add any special handling for this exception (especially for this particular exception type), and can rely on the default handling, unless there is a reason not to do so (e.g. no default handling in this place of code).

Reply:
`getInterserverCredential` exists for InterserverIOHandler mainly. 

My expectation is that credentials should always point to something.

When the server is misconfigured it should fail to start. When the server is misconfigured on reload it should not update the credential (auth bypass). 

With an exception there's a side effect that causes config reload to fail (after updateCredentials). I can catch this in update credentials and skip the update though.

The reason is a shared ptr is to avoid copying or synchronization in InterserverIOHandler. I eluded to this in an earlier comment.